### PR TITLE
Resampler norm

### LINKED ISSUE #1182: Add norm function for NiftyResampler
In order to perform motion compensation with CIL's PDHG, NiftyResampler should have a norm function.

### PR DESCRIPTION
**Pull request by evgueni-ovtchinnikov**
## Changes in this pull request

Implements norm method for NiftyResampler.

## Related issues

Fixes #1182

## Checklist before requesting a review

- [ ] I have performed a self-review of my code
- [ ] I have added docstrings/doxygen in line with the guidance in the developer guide
- [ ] I have implemented unit tests that cover any new or modified functionality
- [x] The code builds and runs on my machine
- [ ] CHANGES.md has been updated with any functionality change

## Contribution Notes

Please read and adhere to the [contribution guidelines](https://github.com/SyneRBI/SIRF/blob/master/CONTRIBUTING.md).

Please tick the following: 

 - [x] The content of this Pull Request (the Contribution) is intentionally submitted for inclusion in SIRF (the Work) under the terms and conditions of the [Apache-2.0](https://www.apache.org/licenses/LICENSE-2.0) License.

### CODE REVIEW

**Comment by KrisThielemans** [comment]
I realise of course this is WIP, but some comments.

Why do we need the `Wrapped_sptr`? I can see it's convenient to have `norm()` etc on a pointer, but why not simply dereference the `shared_ptr`? I guess I'm missing something. 

Also `BFOperator` could be useful I guess, but in its current form, it could just as well be a lambda function probably.

I'm a bit doubtful of including extra classes just for this purpose. If we do, they'd have to be generalised and put in other places once it works.

**Comment by evgueni-ovtchinnikov** [comment]
I just tried to keep modifications to `JacobiCG` and `AcquisitionModel` classes minimal. In my implementation of `JacobiCG`, `vector_type` objects are assumed to have certain methods (`dot` and `axpby`, now also `norm`), and shared pointers do not have them. An alternative to my approach would be to assume all `vector_type` objects to be pointers, but I found it to be too restrictive an assumption about the nature of `vector_type`, and it would require changing all current definitions of `BFOperator` and `norm` for acquisition models, whereas introducing a simple wrapper object for shared pointers allowed me to avoid it.

File: src/Registration/cReg/NiftyResampler.cpp [hunk 1]
```diff
@@ -339,6 +339,28 @@ void NiftyResampler<dataType>::adjoint(std::shared_ptr<ImageData> output_sptr, c
...
+        resampler.set_floating_image(this->floating_image_sptr());
+        resampler.set_interpolation_type(this->get_interpolation_type());
+        auto trans = this->transformations_sptr();
+        for (int i = 0; i < trans.size(); i++)
+	        resampler.add_transformation(this->transformations_sptr()[i]);
+        BFOperator<dataType> bf(sptr_r);
+        JacobiCG<dataType> jcg;
+        jcg.set_num_iterations(num_iter);
+        std::shared_ptr<const ImageData> sptr_im = this->floating_image_sptr();
+        std::shared_ptr<ImageData> sptr_id = sptr_im->clone();
```
Comment (KrisThielemans) [inline src/Registration/cReg/NiftyResampler.cpp:357]:
```suggestion
        auto sptr_id = this->floating_image_sptr()->clone();
```

**Review by KrisThielemans — CHANGES_REQUESTED** [review]
some small suggestions on C++ coding style. However, in the same push please merge `master` back here, such that GitHub Actions can be green :-)

**Pull request merged**